### PR TITLE
Precompute "popular" Records/Results pages in Sidekiq as part of CAD

### PR DESCRIPTION
**Pull request by gregorbg**
Basically to remedy the issue that several threads get fried when people try to access Records pages directly after CAD has been run.

We pre-compute and insert the most popular pages (i.e. the default filter selection that you would see by just opening "Records", but not sophisticated queries like "Female Pyraminx Average History for Guatemala") and insert those into the cache using the same key which the page is expected to fetch just a moment later.

As an additional optimization, we could outsource this new "CAD part" into a separate job so that CAD runtimes aren't skewed. But then the cache timestamp is already invalid and we might not be fast enough with pre-computing until the statistics fetischists out there start bombarding us

### CODE REVIEW

**Comment by gregorbg** [comment]
@thewca-bot deploy staging

**Comment by dunkOnIT** [comment]
I couldn't figure out how to run CAD via the UI, so I did it from the rails console by running `ComputeAuxiliaryData.perform_later()`.

SideKiq shows the job as processed, but the cache keys don't seem to contain results keys: 
```bash
wca-on-rails(prod)* Rails.cache.redis.with do |redis|
wca-on-rails(prod)*   cursor = "0"
wca-on-rails(prod)*   keys = []
wca-on-rails(prod)* 
wca-on-rails(prod)*   loop do
wca-on-rails(prod)*     cursor, batch = redis.scan(cursor)
wca-on-rails(prod)*     keys.concat(batch)
wca-on-rails(prod)*     break if cursor == "0"
wca-on-rails(prod)*   end
wca-on-rails(prod)* 
wca-on-rails(prod)*   keys
wca-on-rails(prod)> end
=> 
["views/competitions/show_results_by_person:66f01f6a0080f11626e4fc075fb29d43/by_person/HuaHinOpen2011/results/query-8798c0314a9ec06884038eabba23c7ed/zh-CN/false",
 "views/competitions/show_results_by_person:66f01f6a0080f11626e4fc075fb29d43/by_person/GajahMadaSpeedcubing2024/results/query-7fe91254b5485588053c76f6ee90aa04/zh-CN/false",
 "views/competitions/show_results_by_person:66f01f6a0080f11626e4fc075fb29d43/by_person/150thCubeMeetinginBrest2017/results/query-7faad227fe980be1454b2013a4f4c6a5/zh-CN/false",
 "views/competitions/show_results_by_person:66f01f6a0080f11626e4fc075fb29d43/by_person/NavidadAvesMaria2022/results/query-8de89a447791e4d031a596d43b922ee0/zh-CN/false",
 "views/competitions/show_results_by_person:66f01f6a0080f11626e4fc075fb29d43/by_person/SanFranciscoBayOpen2017/results/query-3956600736a9a540a122c92097166937/zh-CN/false",
 "views/persons/_results_by_event:624165d9f9937abdc7c3712507e30b41/results_by_event/2024BURU01/2024-12-07 12:39:12 UTC/zh-CN",
 "queues",
 "j|20241217|12:28",
 "processes",
 "views/competitions/show_results_by_person:66f01f6a0080f11626e4fc075fb29d43/by_person/XianNewYear2017/results/query-8bb39d3f111975deb1d43e4809ce5eab/zh-CN/false",
 "views/competitions/show_results_by_person:66f01f6a0080f11626e4fc075fb29d43/by_person/MololoaOpen2024/results/query-b7df900edf4884bc0840dd6fd95dd081/zh-CN/false",
 "views/competitions/show_results_by_person:66f01f6a0080f11626e4fc075fb29d43/by_person/SideNBlindNarreWarren2023/results/query-3b13f4727d7e4f0a43dd0bdc227b9f60/zh-CN/false",
 "views/competitions/show_results_by_person:66f01f6a0080f11626e4fc075fb29d43/by_person/IITKanpurOpen2019/results/query-09da5c8b83e31e30fb2b574490a5269e/zh-CN/false",
 "views/competitions/show_results_by_person:66f01f6a0080f11626e4fc075fb29d43/by_person/1stDibrugarhOpen2018/results/query-a37004e29c458a2cbf1873eaafa0338f/zh-CN/false",
 "stat:processed:2024-12-17",
 "views/competitions/show_results_by_person:66f01f6a0080f11626e4fc075fb29d43/by_person/FinnishChampionship2019/results/query-f7ce6cec24694c7ce9e7404e78a85169/zh-CN/false",
 "stat:failed",
 "views/persons/_results_by_event:624165d9f9937abdc7c3712507e30b41/results_by_event/2018RYBK01/2018-08-06 03:33:53 UTC/zh-CN",
 "ip-172-31-18-24.us-west-2.compute.internal:24:e55ba1c102c9",
 "views/competitions/show_results_by_person:66f01f6a0080f11626e4fc075fb29d43/by_person/Israel2011/results/query-a834aa75b776aa5ce621dc67919a2fb2/zh-CN/false",
 "stat:processed",
 "stat:failed:2024-12-17",
 "views/competitions/show_results_by_person:66f01f6a0080f11626e4fc075fb29d43/by_person/AvenidaChileVII2024/results/query-fb9a5b750a5a619020061a78b963474d/zh-CN/false",
 "views/competitions/show_results_by_person:66f01f6a0080f11626e4fc075fb29d43/by_person/KharkivOpen2018/results/query-6f74262e85d2f0df8b5b933b6510dc7e/zh-CN/false",
 "ComputeAuxiliaryData-17-12:28",
 "views/competitions/show_results_by_person:66f01f6a0080f11626e4fc075fb29d43/by_person/CubeWithIntentioninDenton2022/results/query-1ccfc24872330af14b9dcdebdf2b9ebd/zh-CN/false",
 "views/competitions/show_results_by_person:66f01f6a0080f11626e4fc075fb29d43/by_person/EuskadiOpenPortugalete2018/results/query-8770bb9ce5148c01d0738265c6c3334a/zh-CN/false",
 "views/competitions/show_results_by_person:66f01f6a0080f11626e4fc075fb29d43/by_person/ComodoroCubea2023/results/query-af30cc4a2ca6993132d16c52fc22ce0d/zh-CN/false",
 "views/persons/_results_by_event:624165d9f9937abdc7c3712507e30b41/results_by_event/2024BRET03/2024-07-29 04:25:52 UTC/zh-CN",
 "views/competitions/show_results_by_person:66f01f6a0080f11626e4fc075fb29d43/by_person/ManufakturaOpen2013/results/query-7589818f6c7555526f40785a8158bd6b/zh-CN/false",
 "views/competitions/show_results_by_person:66f01f6a0080f11626e4fc075fb29d43/by_person/MasterpressSummer2014/results/query-a8021f17028b157e8f994f13e6d38709/zh-CN/false"]
```

(Command is a ChatGPT suggestion to get around the complexities of timeouts and a Redis connection pool - I used it before clearing cache, and it seems to work as intended?)

**Comment by dunkOnIT** [comment]
I tried to read the contents of a few of the cache keys which seemed interesting, but none yielded particularly useful results: 
```bash
wca-on-rails(prod)> Rails.cache.read('stat:failed')
=> nil
wca-on-rails(prod)> Rails.cache.read('stat:failed:2024-12-17')
=> nil
wca-on-rails(prod)> Rails.cache.read('ComputeAuxiliaryData-17-12:28')
/rails/vendor/bundle/ruby/3.3.0/gems/irb-1.14.1/lib/irb.rb:1260:in `full_message': incompatible marshal file format (can't be read) (TypeError)
	format version 4.8 required; 0.0 given
wca-on-rails(prod)* Rails.cache.redis.with do |redis|
wca-on-rails(prod)*   redis.get("ComputeAuxiliaryData-17-12:28")
wca-on-rails(prod)> end
=> "\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0001"
wca-on-rails(prod)> 
```

Re the `ComputeAuxiliaryData-17-12:28` key, ChatGPT suggested the new command because of the following: 

> The error indicates that the cached data at the specified key was serialized using a different version of Ruby's Marshal format, and the current environment cannot interpret it correctly. This typically happens when:
> 
>     The data was cached with a different Ruby version that uses an incompatible Marshal format.
>     Non-Marshal data was stored at the key but is being read as if it was Marshaled.

**Comment by maxidragon** [comment]
> I couldn't figure out how to run CAD via the UI, so I did it from the rails console by running `ComputeAuxiliaryData.perform_later()`.

You can just visit `/admin/compute_auxiliary_data` and click "Do it!"

File: app/jobs/compute_auxiliary_data.rb [hunk 1]
```diff
@@ -10,4 +10,138 @@ def self.reason_not_to_run
   def perform
     AuxiliaryDataComputation.compute_everything
   end
+
+  def recompute_popular_rankings
```
Comment (dunkOnIT) [inline app/jobs/compute_auxiliary_data.rb:14]: I'm not seeing how this is called/invoked - AuxiliaryDataComputation doesn't invoke it, and I don't see any hooks to run it after CAD has been completed. Grepping for the method name only yields the method definition

Comment (gregorbg) [reply app/jobs/compute_auxiliary_data.rb:18]: Uhm... Look again please? :innocent: 

(Yes, this was my fault and I just pushed a commit to fix it...)

**Comment by dunkOnIT** [comment]
@thewca-bot deploy staging

**Comment by dunkOnIT** [comment]
@thewca-bot deploy staging

**Comment by dunkOnIT** [comment]
@thewca-bot deploy staging

**Comment by dunkOnIT** [comment]
333 rankings took a long time (~30s per network tab) to load, despite the cache key appearing to be in the redis cache (see pic 2)
image

image

NewRelic confirms that we appear to be recomputing these pages, not hitting the cache for them. However, once they've been hit once, they calculate quickly.
image

**Comment by dunkOnIT** [comment]
Here are all the cache keys after I've hit those pages - trying to see 

```
["records/555/world/all years/All/2024-12-30 14:35:46 UTC",
 "records/minx/world/all years/All/2024-12-30 14:35:46 UTC",
 "rankings/pyram/world/all years/All/average/2024-12-30 14:35:46 UTC",
 "rankings/clock/world/all years/All/average/2024-12-30 14:35:46 UTC",
 "records/333fm/world/all years/All/2024-12-30 14:35:46 UTC",
 "views/competitions/show_results_by_person:66f01f6a0080f11626e4fc075fb29d43/by_person/Launceston2022/results/query-c2899d20461b2b1ea19768112454b80d/zh-CN/false",
 "views/competitions/show_results_by_person:66f01f6a0080f11626e4fc075fb29d43/by_person/CostaRicaNationals2018/results/query-3215f16ace3e89edebf18ddaad1cee36/zh-CN/false
",
 "rankings/444/world/all years/All/average/2024-12-30 14:35:46 UTC",
 "views/results/records:136ddad6718bfd828a735b214ec1106e/records/all events/world/all years/mixed/All/2024-12-30 14:35:46 UTC/en",
 "views/competitions/show_results_by_person:66f01f6a0080f11626e4fc075fb29d43/by_person/BayAreaSpeedcubin102017/results/query-e9e126fd3656416bea46cb16ea26010a/zh-CN/false
",
 "records/pyram/world/all years/All/2024-12-30 14:35:46 UTC",
 "rankings/minx/world/all years/All/average/2024-12-30 14:35:46 UTC",
 "views/competitions/show_results_by_person:66f01f6a0080f11626e4fc075fb29d43/by_person/VancouverSummer2012/results/query-7391832c88d31fe27a847f30a3185da8/zh-CN/false"
,
 "rankings/333bf/world/all years/All/single/2024-12-30 14:35:46 UTC",
 "views/competitions/show_results_by_person:66f01f6a0080f11626e4fc075fb29d43/by_person/TrujilloHappyNewYear2020/results/query-0fb0cb458bbe1cbe83fe0ff3169bf305/zh-CN/false
",
 "views/competitions/show_results_by_person:66f01f6a0080f11626e4fc075fb29d43/by_person/LittleGiantOpen2023/results/query-ea12b5647cdb87c122d9c0de7fe57037/zh-CN/false"
,
 "views/competitions/show_results_by_person:66f01f6a0080f11626e4fc075fb29d43/by_person/PDPUOpen2019/results/query-a9c87c199c9690092406ec1842323aeb/zh-CN/false",
 "views/competitions/show_results_by_person:66f01f6a0080f11626e4fc075fb29d43/by_person/UlaanbaatarCubingFeb2024/results/query-f544804220857eb3847ce27462d4767a/zh-CN/false
",
 "rankings/222/world/all years/All/single/2024-12-30 14:35:46 UTC",
 "views/competitions/show_results_by_person:66f01f6a0080f11626e4fc075fb29d43/by_person/RialtoCubingSummer2023/results/query-4ecf8173bedfd3fee0cb74000d8bad8f/zh-CN/false
",
 "rankings/444bf/world/all years/All/average/2024-12-30 14:35:46 UTC",
 "views/competitions/show_results_by_person:66f01f6a0080f11626e4fc075fb29d43/by_person/NortheasternFall2015/results/query-7524e1d6c40d747ec84bd542e0513b61/zh-CN/false"
,
 "rankings/333mbf/world/all years/All/single/2024-12-30 14:35:46 UTC",
 "records/333mbf/world/all years/All/2024-12-30 14:35:46 UTC",
 "rankings/444/world/all years/All/single/2024-12-30 14:35:46 UTC",
 "views/competitions/show_results_by_person:66f01f6a0080f11626e4fc075fb29d43/by_person/SnoCoSummer2018/results/query-604e083102770994b615b9633f7130e9/zh-CN/false",
 "views/competitions/show_results_by_person:66f01f6a0080f11626e4fc075fb29d43/by_person/PuneUnscrambled2023/results/query-4587f94ca43e6776814b9bf951c63952/zh-CN/false"
,
 "rankings/333fm/world/all years/All/single/2024-12-30 14:35:46 UTC",
 "records/clock/world/all years/All/2024-12-30 14:35:46 UTC",
 "views/competitions/show_results_by_person:66f01f6a0080f11626e4fc075fb29d43/by_person/XuzhouSpring2019/results/query-69f6567ed7bfde3e3019f6b65bf9819b/zh-CN/false",
 "rankings/sq1/world/all years/All/average/2024-12-30 14:35:46 UTC",
 "views/competitions/show_results_by_person:66f01f6a0080f11626e4fc075fb29d43/by_person/SaratogaSpringsSummer2024/results/query-64ae9758ca5d45aac6dcfaee45a2bb2b/zh-CN/false
",
 "rankings/777/world/all years/All/single/2024-12-30 14:35:46 UTC",
 "rankings/333mbf/world/all years/All/average/2024-12-30 14:35:46 UTC",
 "views/competitions/show_results_by_person:66f01f6a0080f11626e4fc075fb29d43/by_person/SuzhouOpen2014/results/query-953817a0be186503af4a09fa55d38ed8/zh-CN/false",
 "views/competitions/show_results_by_person:66f01f6a0080f11626e4fc075fb29d43/by_person/SiouxCitySunday2023/results/query-63f613a80eb277d1331b19fdb48d2113/zh-CN/false"
,
 "rankings/555/world/all years/All/single/2024-12-30 14:35:46 UTC",
 "records/666/world/all years/All/2024-12-30 14:35:46 UTC",
 "rankings/333/world/all years/All/average/2024-12-30 14:35:46 UTC",
 "records/333oh/world/all years/All/2024-12-30 14:35:46 UTC",
 "records/all events/world/all years/mixed/All/2024-12-30 14:35:46 UTC",
 "views/competitions/show_results_by_person:66f01f6a0080f11626e4fc075fb29d43/by_person/TurkishSummerCup2013/results/query-ceb2fbb5b70ebbf3627c6e3b8fe38f7d/zh-CN/false"
,
 "rankings/222/world/all years/All/average/2024-12-30 14:35:46 UTC",

**Comment by gregorbg** [comment]
> Edit: Here are the 3 cache keys that deal with 333 single - the first one looks suspicious to me, I wonder if _that_ is the cached page that's generated once the page is visited. I'll investigate tomorrow, quite a long day today.

No, it's not the first one. It's a subtle difference between the second and third one. Thank you for your investigation though! It led me right to the point: You will see that the longer, third key contains a `100 persons` tag string, which is automatically set by the view controller code as a "default value". That means, the public page upon access tries to load the data including this cache key.

The new CAD code which I'm writing as part of this PR though did not include this part. I have changed it and pushed a fresh commit, which should now (hopefully, third time's the charm!) correctly populate the cache using a key that the controller can find.

(In case you're interested, the "obscure" first key is an additional piece of caching we're doing: First, we cache the raw SQL result under the more general, kinda-readable humanized key, and then we _also_ cache the HTML code that renders the tables under a separate, Rails-managed, hashed key.)

**Review by dunkOnIT — APPROVED** [review]
It works! And thanks for the explanation re the keys :D

**Pull request merged**